jsonrpc/server.go

```diff
@@ -211,59 +216,78 @@ func (s *Server) HandleReader(ctx context.Context, reader io.Reader) ([]byte, er
 		}
 	} else {
 		var batchReq []json.RawMessage
-		var batchRes []json.RawMessage
 
 		if batchJSONErr := dec.Decode(&batchReq); batchJSONErr != nil {
 			res.Error = Err(InvalidJSON, batchJSONErr.Error())
 		} else if len(batchReq) == 0 {
 			res.Error = Err(InvalidRequest, "empty batch")
 		} else {
-			for _, rawReq := range batchReq { // todo: handle async
-				var resObject *response
+			return s.handleBatchRequest(ctx, batchReq)
+		}
+	}
 
-				reqDec := json.NewDecoder(bytes.NewBuffer(rawReq))
-				reqDec.UseNumber()
+	if res == nil {
+		return nil, nil
+	}
+	return json.Marshal(res)
+}
 
-				req := new(request)
-				if jsonErr := reqDec.Decode(req); jsonErr != nil {
+func (s *Server) handleBatchRequest(ctx context.Context, batchReq []json.RawMessage) ([]byte, error) {
+	var batchRes []json.RawMessage
+	var resMutex sync.Mutex
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var wg sync.WaitGroup
+	wg.Add(len(batchReq))
+	for _, rawReq := range batchReq {
+		rawReq := rawReq
+		s.pool.Go(func() {
+			defer wg.Done()
+			var resObject *response
+
+			reqDec := json.NewDecoder(bytes.NewBuffer(rawReq))
+			reqDec.UseNumber()
+
+			req := new(request)
+			if jsonErr := reqDec.Decode(req); jsonErr != nil {
+				resObject = &response{
+					Version: "2.0",
+					Error:   Err(InvalidRequest, jsonErr.Error()),
+				}
+			} else {
```

Comment:
Nit; we could do this outside of go routine, to avoid spinning up new go routines for obviously invalid requests.

Reply:
Rewrote it